primitives/npos-elections/compact/tests/compact.rs

```diff
@@ -0,0 +1,168 @@
+use std::{
+	collections::{HashSet, HashMap},
+	convert::TryInto,
+	hash::Hash,
+};
+use sp_npos_elections::{CompactSolution, IndexAssignment};
+use rand::{self, Rng, SeedableRng, seq::SliceRandom};
+
+const CANDIDATES: usize = 1000;
+const VOTERS: usize = 2500;
+
+sp_npos_elections_compact::generate_solution_type!(
+	#[compact]
+	pub struct Compact::<VoterIndex = u32, TargetIndex = u16, Accuracy = Accuracy>(16)
+);
+
+pub type AccountId = u64;
+/// The candidate mask allows easy disambiguation between voters and candidates: accounts
+/// for which this bit is set are candidates, and without it, are voters.
+pub const CANDIDATE_MASK: AccountId = 1 << ((std::mem::size_of::<AccountId>() * 8) - 1);
+pub type CandidateId = AccountId;
+
+pub type Accuracy = sp_runtime::Perbill;
+
+pub type Assignment = sp_npos_elections::Assignment<AccountId, Accuracy>;
+pub type Voter = (AccountId, sp_npos_elections::VoteWeight, Vec<AccountId>);
+
+/// Generate voter and assignment lists. Makes no attempt to be realistic about winner or assignment fairness.
+///
+/// Maintains these invariants:
+///
+/// - candidate ids have `CANDIDATE_MASK` bit set
+/// - voter ids do not have `CANDIDATE_MASK` bit set
+/// - assignments have the same ordering as voters
+/// - `assignments.distribution.iter().map(|(_, frac)| frac).sum() == One::one()`
+/// - a coherent set of winners is chosen.
+/// - the winner set is a subset of the candidate set.
+/// - `assignments.distribution.iter().all(|(who, _)| winners.contains(who))`
+fn generate_random_votes(
+	candidate_count: usize,
+	voter_count: usize,
+	mut rng: impl Rng,
+) -> (Vec<Voter>, Vec<Assignment>, Vec<CandidateId>) {
```

Comment:
a function like this already exists somewhere else in the codebase no?

@kianenigma

Reply:
The closest I've seen are the functions in [`common.rs`](https://github.com/paritytech/substrate/blob/master/primitives/npos-elections/fuzzer/src/common.rs), and those don't quite do what I need.

Reply:
yeah unfortunately they are different. 

Note that we also had similar function in staking and we removed them. 

A good course of action would be to pull `common.rs` outside of the fuzzer and put this also next to them, if possible. If too much of a PITA for whatever reason, fine for now.

Reply:
> pull common.rs outside of the fuzzer 

Yup: https://github.com/paritytech/substrate/pull/8614/commits/eed1c218823794d1977e6e1ac1f4f70b3aa5a6e9